doc/specs/#2046 - Unified keybindings and commands, and synthesized action names.md

```diff
@@ -0,0 +1,589 @@
+the text for the action in the Action Mode of the Command Palette. However, this
+is a little tedious for users who already have lots of keys bound. They'll need
+to go through and add names to each of their existing keybindings to ensure that
+the actions appear in the palette. Could we instead synthesize the names for the
+commands ourselves? This  would enable users to automatically get each of their
+existing keybindings to appear in the palette without any extra work.
+
+To support this, the following changes will be made:
+* `ActionAndArgs` will get a `GenerateName()` method defined. This will create a
+  string describing the `ShortcutAction` and it's associated `ActionArgs`.
+  - Not EVERY action _needs_ to define a result for `GenerateName`. Actions that
+    don't _won't_ be automatically added to the Command Palette.
+  - Each of the strings used in `GenerateName` will need to come from our
+    resources, so they can be localized appropriately.
+* When we're parsing commands, if a command doesn't have a `name`, we'll instead
+  attempt to use `GenerateName` to create the unique string for the action
+  associated with this command. If the command does have a `name` set, we'll use
+  that string instead, allowing the user to override the default name.
+  - If a command has it's name set to `null`, then we'll ignore the command
+    entirely, not just use the generated name.
+
+[**Appendix 1**](#appendix-1-name-generation-samples-for-ShortcutActions) below
+shows a complete sample of the strings that will be generated for each of the existing
+`ShortcutActions`, and many of the actions that have been proposed, but not yet
+implemented.
+
+These strings should be human-friendly versions of the actions and their
+associated args. For some of these actions, with very few arguments, the strings
+can be relatively simple. Take for example, `CopyText`:
+
+JSON | Generated String
+-- | --
+`{ "action":"copyText" }` | "Copy text"
+`{ "action":"copyText", "singleLine": true }` | "Copy text as a single line"
+`{ "action":"copyText", "singleLine": true, "copyFormatting": true }` | "Copy text without formatting"
+`{ "action":"copyText", "singleLine": true, "copyFormatting": true }` | "Copy text as a single line without formatting"
+
+CopyText is a bit of a simplistic case however, with very few args or
+permutations of argument values. For things like `newTab`, `splitPane`, where
+there are many possible arguments and values, it will be acceptable to simply
+append `", property:value"` strings to the generated names for each of the set
+values.
+
+For example:
+
+JSON | Generated String
+-- | --
+`{ "action":"newTab", "profile": "Hello" }` | "Open a new tab, profile:Hello"
+`{ "action":"newTab", "profile": "Hello", "directory":"C:\\", "commandline": "wsl.exe", title": "Foo" }` | "Open a new tab, profile:Hello, directory:C:\\, commandline:wsl.exe, title:Foo"
+
+
+This is being chosen in favor of something that might be more human-friendly,

```

Comment:
Consider the following...
```js
{ "command": { "action": "newTab", "index": 0 }, "keys": "ctrl+a" },
{ "command": { "action": "newTab", "profile": <profile-at-index-0> }, "keys": "ctrl+b" },
{ "command": { "action": "newTab", "index": 0 }, "keys": "ctrl+c" },
``` 

First, I think we should resolve what `"index": 0` means when we generate a name for these commands. Most people don't think "I want to open my 3rd profile". That's not descriptive/helpful. We should generate "open a new tab with profile <name>". When https://github.com/microsoft/terminal/pull/5888 goes in, this should resolve to whatever the target actually is.

Second, if we go with the above plan, I think in this situation, we should only register one string to the Command Palette. Since "ctrl+c" is the last one defined, "ctrl+c" is shown in the Command Palette.

Third, if a user then adds this to their settings.json:
```js
{ "name": "foo", "command": { "action": "newTab", "index": 0 } },
```
We just register a new "name" and don't touch the pre-registered ones?

I've confused myself but I think you can follow where I'm concerned.

Reply:
Yea, #5888 is part of the complication, which is why I'm thinking we should change that part of the spec back to "open the Nth profile". I really _really_ don't like the idea of the `newTab(index=i)` action needing to know the entire structure of the new tab dropdown so that it could generate a name for itself.

Reply:
Second, I'm thinking that this should be two commands in the palette:
```
Open new tab, profile: <profile and index 0>[ctrl+b]
Open new tab, profile index: 0 [ctrl+a], [ctrl+c]
```

I'm not sold on the "display _every_ keybinding for an action in the command palette" thing quite yet, I'm pretty sure that 

```
Open new tab, profile: <profile and index 0>[ctrl+b]
Open new tab, profile index: 0 [ctrl+a]
```
would be acceptable for now.

<hr>

Third, yep, then the command palette would be 

```
Open new tab, profile: <profile and index 0>[ctrl+b]
Open new tab, profile index: 0 [ctrl+a]
foo
```